doc/source/development/rfc/rfc100_float16_support.rst

```diff
@@ -0,0 +1,142 @@
+.. _rfc-100:
+
+=============================
+RFC 100: Support float16 type
+=============================
+
+============== =============================================
+Author:        Erik Schnetter
+Contact:       eschnetter @ perimeterinstitute.ca
+Started:       2024-Jun-05
+Status:        Proposed
+Target:        GDAL 3.11
+============== =============================================
+
+Summary
+-------
+
+This RFC adds support for the IEEE 16-bit floating point data type
+(aka ``half``, ``float16``). It adds new pixel data types
+``GDT_Float16`` and ``GDT_CFloat16``, backed by a new ``GFloat16`` C++
+type. This type will be emulated in software if not available
+natively.
+
+Some drivers, in particular Zarr, will be extended to support this
+type. Drivers that do not support this type will be checked to ensure
+they report an appropriate errors when necessary.
+
+Motivation
+----------
+
+- The float16 type is increasingly common. It is supported by GPUs, by
+  modern CPUs, and by modern compilers (e.g. GCC 12).
+
+- Using ``uint16`` values to shuffle ``float16`` values around is
+  inconvenient and awkward. There is no mechanism in GDAL to mark
+  attributes or datasets which ``uint16`` values should be interpreted
+  as ``float16``.
+
+- Some drivers (at least the HDF5, GTiff, and Zarr) already handle
+  float16 by exposing it as float32, using software conversion
+  routines. This type should be supported without converting to
+  float32.
+
+- C++23 will introduce native support for ``std::float16_t``. However,
+  it will likely be several years until C++23 will be a requirement
```

Comment:
What is the transition plan for when C++23 is the minimum for GDAL? For those who have C++23, can they use `std::float16_t`?

Reply:
That's a good idea. I'll add that.

As a side note, using `_Float16` directly (that type is widely supported) was not working because e.g. `std::numeric_limits<_Float16>` was not working, and neither was `std::isnan(_Float16)`. Using `std::float16_t` directly will only work if the C++ standard library supports it.